### DISCUSSION TITLE
How to let a point follow a motion path to move

### DISCUSSION THREAD

**Discussion by kittykisser** [sofa-framework/sofa#4143] [Develop in C++] [answered]
![900925a816ad1caf183845819f46f2c](https://github.com/sofa-framework/sofa/assets/81600464/9e2551e3-69b2-4668-91c3-f5ca7e1b83cb)

This is a simulation I built, with a soft rope and a wall. I want to simulate the process of swinging the soft rope and then the tail of rope will hit on the wall. However, I do not know how to apply a motion track to the head of the soft rope, I hope that some partners can give me some help. This is my python source code for building this model.

```
# -*- coding: utf-8 -*-
from stlib.physics.deformable import ElasticMaterialObject
from stlib.physics.constraints import FixedBox
from stlib.physics.collision import CollisionMesh
from stlib.scene import Node 
from softrobots.actuators import PullingCable
from splib.loaders import loadPointListFromFile

import Sofa


def String(parentNode=None, name="String",
           rotation=[0.0, 0.0, 0.0], translation=[0.0, 0.0, 0.0],
           fixingBox=[0.0,0.0,0.0], pullPointLocation=[0.0,0.0,0.0]):

    string = Node(parentNode, "String")
    eobject = ElasticMaterialObject(string,
                                    volumeMeshFileName="data/string.vtk",
                                    surfaceMeshFileName="data/string.stl",
                                    surfaceColor=[0.0,0.7,0.8]
                                    )
   
    CollisionMesh(eobject,
         surfaceMeshFileName="data/string.stl", name="part0")



def createScene(rootNode):
    # -*- coding: utf-8 -*-
    from stlib.scene import MainHeader, ContactHeader
    from stlib.visuals import ShowGrid
    from stlib.physics.rigid import Floor
    from stlib.physics.rigid import Cube
    m=MainHeader(rootNode, plugins=["SoftRobots"],gravity=[0.0,-981.0,0.0])
    m.getObject("VisualStyle").displayFlags='showForceFields showBehaviorModels showInteractionForceFields showCollisionModels'
    ContactHeader(rootNode, alarmDistance=4, contactDistance=3, frictionCoef=0.08)
    Floor(rootNode,
          translation=[0.0,0.0,-500.0],
          uniformScale=5.0,
          rotation=[90, 0, 0],
          isAStaticObject=True)

    #ShowGrid(rootNode)
    String(rootNode)

    return rootNode
```

**Comment by bakpaul** [comment, marked as answer]
Hi, for this I suggest to use the component `LinearMovementConstraint`. It is a projective constraint forcing the value of the position of a point following a parametric linear motion. Thank's to that you can enforce the position of a point inside of a isolated node and use it to move the base of your rope by linking its base to this point with a `RestShapeSpringsForceField`. e.g. : 
```
<Node name="MotionNode">
          <MechanicalObject name="motion" template="Rigid3d" position="0 0 0 0 0 0 1" showObject="true" showObjectScale="0.01" />
          <LinearMovementConstraint 
              name='Movement' 
              keyTimes='0 0.1 20 25 30 35 40 45 50 55 60' 
              movements='
                      0 0 0 0 0 0  
                      0 0 0 0 0 0  
                      20 0 0 0 0 0  
                      20 8 0 0 0 0 
                      20 0 0 0 0 0
                      20 -8 0 0 0 0
                      20 0 0 0 0 0
                      20 0 4 0 0 0
                      20 0 0 0 0 0
                      20 0 -4 0 0 0
                      20 0 0 0 0 0'/>
</Node>

<Node name="Rope">
      ....
      <RestShapeSpringsForceField external_rest_shape="@../MotionNode/motion" stiffness="10000000"  angularStiffness="10000000" points="0" external_points="0" />
       ...
<Node/>
```

My example is in xml but can be easily translated in python. 

Using a spring forcefield allow better performance in term of convergence of the Lagrangian constraint solving. But the tradeoff is that the rope base will not be exactly at the point defined by the 'MotionNode' and will not follow exactly the orientation. If you want more rigidity, you should use langrangian based constraint to fix the base (see `BilateralInteractionConstraint`)

**Reply by bakpaul** [reply]
Did that work for you @kittykisser ?

**Reply by kittykisser** [reply]
![string_00000001](https://github.com/sofa-framework/sofa/assets/81600464/4e12ae9d-6e27-4f38-9b43-0c7f46bfca86)

Yes, I have nailed this problem! Sorry to make a late reply for you. Thank you so much! And I have another little question. I want to stop the animation when the tail of the soft rope hit the wall. How can I make this come true. The logic I imagine is to stop the animation when the tail coordinates are equal to the wall position coordinates and get the coordinates of the hit-position.

**Reply by bakpaul** [reply]
Fro this I suggest you to look at https://github.com/sofa-framework/sofa/discussions/4161
